src/hosting/config.ts

```diff
@@ -0,0 +1,256 @@
+import path from "path";
+
+// TODO: Consider putting this type along with the Options type.
+// I haven't tried to do this yet because true options embeds classes, not just
+// interfaces.
+// We should consider either refactoring options so it can be more easily mocked,
+// or creating a utility for constructing options easily in tests.
+export interface MockableOptions {
+  project?: string;
+  site?: string;
+  config: {
+    src: FirebaseConfig;
+  };
+  rc: {
+    requireTarget(project: string, type: string, name: string): string[];
+  };
+  cwd?: string;
+  configPath?: string;
+  only?: string;
+  except?: string;
+  normalizedHostingConfig?: HostingMultiple;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+const mockableOptionsIsCompatibleWithOptions: Implements<Options, MockableOptions> = true;
+
+// assertMatches allows us to throw when an --only flag doesn't match a target
+// but an --except flag doesn't. Is this desirable behavior?
+function matchingConfigs(
+  configs: HostingMultiple,
+  targets: string[],
+  assertMatches: boolean
+): HostingMultiple {
+  const matches: HostingMultiple = [];
+  const [hasSite, hasTarget] = partition(configs, (c) => "site" in c);
+  for (const target of targets) {
+    const siteMatch = hasSite.find((c) => c.site === target);
+    const targetMatch = hasTarget.find((c) => c.target === target);
+    if (siteMatch) {
+      matches.push(siteMatch);
+    } else if (targetMatch) {
+      matches.push(targetMatch);
+    } else if (assertMatches) {
+      throw new FirebaseError(
+        `Hosting site or target ${bold(target)} not detected in firebase.json`
+      );
+    }
+  }
+  return matches;
+}
+
+/**
+ * Returns a subset of configs that match the only string
+ */
+export function filterOnly(configs: HostingMultiple, onlyString?: string): HostingMultiple {
+  if (!onlyString) {
+    return configs;
+  }
+
+  let onlyTargets = onlyString.split(",");
+  // If an unqualified "hosting" is in the --only,
+  // all hosting sites should be deployed.
+  if (onlyTargets.includes("hosting")) {
+    return configs;
+  }
+
+  // Strip out Hosting deploy targets from onlyTarget
+  onlyTargets = onlyTargets
+    .filter((target) => target.startsWith("hosting:"))
+    .map((target) => target.replace("hosting:", ""));
+
+  return matchingConfigs(configs, onlyTargets, /* assertMatch= */ true);
+}
+
+/**
+ * Returns a subset of configs that match the except string;
+ */
+export function filterExcept(configs: HostingMultiple, exceptOption?: string): HostingMultiple {
+  if (!exceptOption) {
+    return configs;
+  }
+
+  const exceptTargets = exceptOption.split(",");
+  if (exceptTargets.includes("hosting")) {
+    return [];
+  }
+
+  const exceptValues = exceptTargets
+    .filter((t) => t.startsWith("hosting:"))
+    .map((t) => t.replace("hosting:", ""));
+  const toReject = matchingConfigs(configs, exceptValues, /* assertMatch=*/ false);
```

Comment:
Is that comment just a stray from development? I can't tell what it's trying to say.